CSV analysis module

AdaraPutri opened the pull request:
## 📝 Description

Initially the purpose of this PR was just to add support for CSV files in the llm text analysis pipeline. After addressing requested changes, it now adds a new separate CSV analysis module and refactors some extension checks in helpers.py to refer to the database.

Key changes:

- Added csv_analyze.py for independent dataset analysis (works for both standalone CSV projects and supporting CSVs)
- Updated project_analysis.py to correctly route CSV files to the new module
- Refactored helpers.py and related modules to fetch file extensions from the database instead of from file paths
- Updated text_llm_analyze.py and alt_analyze.py and their tests to include DB connections for consistency
- Added tests for the new csv module

Now, when you upload just a csv file, or when a text file project includes csv files as part of data collection the terminal includes these metadata (on top of the text file analysis if present):

- Number of rows, columns, headers, and data types
- Missing row counts and percentages
- Numeric column summaries (min, max, mean)
- User-typed or llm-generated dataset summary
- Growth trend tracking for multi-version datasets (e.g., data.csv, data2.csv, data3.csv)

**Closes:** #178 

---

## 🔧 Type of Change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation added/updated
- [x] ✅ Test added/updated
- [x] ♻️ Refactoring
- [ ] ⚡ Performance improvement

---

## 🧪 Testing

- [x] Verified that all tests pass when python -m pytest is run
- [x] Manually tested with project file that contains one csv dataset along with text files
- [x] Manually tested with project file that contains multiple csv datasets (different versions) along with text files
- [x] Manually tested with project file that contains ONLY a csv dataset

---

## ✓ Checklist

- [x] 🤖 GenAI was used in generating the code and I have performed a self-review of my own code
- [x] 💬 I have commented my code where needed
- [ ] 📖 I have made corresponding changes to the documentation
- [x] ⚠️ My changes generate no new warnings
- [x] ✅ I have added tests that prove my fix is effective or that my feature works and tests are passing locally
- [x] 🔗 Any dependent changes have been merged and published in downstream modules
- [ ] 📱 Any UI changes have been checked to work on desktop, tablet, and/or mobile

---

## 📸 Screenshots

When llm consent is given, project has multiple versions of csv dataset AND text files:
<img width="950" height="723" alt="Screenshot 2025-11-09 at 3 58 18 PM" src="https://github.com/user-attachments/assets/491864b4-5c03-495a-a4d1-94a46a040042" />
<img width="947" height="781" alt="Screenshot 2025-11-09 at 3 58 29 PM" src="https://github.com/user-attachments/assets/6d805ac6-5fe2-4c88-b190-9dd70e1c5516" />
<img width="945" height="762" alt="Screenshot 2025-11-09 at 3 58 40 PM" src="https://github.com/user-attachments/assets/8e11c1b8-3ad8-421e-b488-bee6b3e4f7f1" />

When llm consent is not given, project is just a single csv file:
<img width="949" height="681" alt="Screenshot 2025-11-09 at 4 00 22 PM" src="https://github.com/user-attachments/assets/20f03baa-c765-4780-bbf6-d62c22c91a68" />

AdaraPutri committed:
added an extractfromcsv module in helpers.py

AdaraPutri committed:
modified llm prompts in skills and success factors to include threshold for analyzing csv metadata

AdaraPutri committed:
fixed llm prompt

AdaraPutri committed:
ensure all tests passed

ammaarkhan reviewed (changes requested):
Hey Adara, great work on this! Adding CSV for the text analysis is a great add on to the existing functionality.

This was brought up in a previous PR #160, when CSV was listed as a file type for text analysis but selecting it wouldn't display any analysis. I tested this scenario and the CSV was analyzed and the output was as desired. Great work!
<img width="691" height="602" alt="Image" src="https://github.com/user-attachments/assets/53e077f5-4976-4dd5-a06d-6db7a8fcaa4d" />

All tests also passed:
<img width="829" height="482" alt="Image" src="https://github.com/user-attachments/assets/b30e9287-281f-4c0c-95a6-161cb8936d20" />

However, I noticed there's no unit test for the new extractfromcsv() function, and no integration test showing how CSV metadata is handled in the LLM pipeline. Could you add test coverage for these scenarios? This will help ensure the CSV functionality doesn't break in future changes.

AdaraPutri committed:
Merge remote-tracking branch 'origin/main' into llm-text-analysis

AdaraPutri committed:
added test cases for when csv files are in supporting and not, tests if largest file auto selects and supporting file detected flows correctly

AdaraPutri committed:
ensured tests passed

taoTimTim reviewed (changes requested):
Some good work here, but I see a few errors and duplications that need to be addressed.

Inside `helpers.py` you are extracting the file extension from the file path. This should be changed to call the extension from the database, I would create a helper. Checking if it is valid is a duplication, there is a reason we checked the file extension inside `parsing.db`, so we would not have to later. 

When I ran the code with JUST a .csv file inside the project, the terminal closed with an error:
<img width="924" height="559" alt="image" src="https://github.com/user-attachments/assets/1353d20e-83e0-4c56-8d91-6340b5327f06" />

This is because the csv file is not actually a text file, meaning it does not contain text, it is data, and `analyze_linguistic_complexity()` expects a string. We store it as text but this may be a design flaw.

There are three solutions to this:
1.  Change the system architecture to detect data extensions instead of just text and code, that way it will have a different flow later on.
2. Check the return type before analysis, if it is a .csv file we analyze it differently.
3. Return an error to the user when there is just a .csv file in the project, and inform them it is not a valid project

I personally think number 3 is an easy way out, because it is possible for a project to be data oriented, and thus have only .csv files, so I do not think number 3 is the solution. 

Lastly, I'm not sure how this kind of data returned from a .csv file is useful for the end goal (portfolios and resumes). Use the LLM to figure out what the data represents, what it is, etc., not just that there is little data. I inputted a .csv file with only a few lines, and it said that there is missing data. This is false, there are cases where a .csv file WILL have only a little bit of data. 

We should be figuring out what the user did with the data. While metadata is cool to look at, it is not helpful later on when we generate the resumes and portfolios.

Good work here, but needs some changes.

ivonanicetin reviewed (changes requested):
Good work I think expanding to extend to allow csv files is useful.

I ran into a few issues, when a csv file is the main file this causes errors I believe this is because main_text becomes a dict not a string but analyze_linguistic_complexity() expects a string. I've added screenshot below. 

I also think it is not the most insightful or just if we see the missing data as a weakness since I think there are many projects that have empty elements. I think it would be more insightful and useful for later if we use the LLM to help infer the actual project based on the header, the data itself and also maybe getting the means , min/max from the columns so we can better analysis later for when we generate the portfolio/resume.

<img width="1404" height="762" alt="Screenshot 2025-11-08 011151" src="https://github.com/user-attachments/assets/27f828a5-d6a2-4408-9082-a4e517e6cc06" />
<img width="1413" height="279" alt="Screenshot 2025-11-08 011206" src="https://github.com/user-attachments/assets/b87bdb11-348b-4ade-9e6c-b85a9e068e65" />

AdaraPutri committed:
Merge remote-tracking branch 'origin/main' into llm-text-analysis

AdaraPutri committed:
refactored helpers.py to remove manual extension validation and fetch file extension from DB

AdaraPutri committed:
update relevant helper calls to include DB connection and revert text_llm_analyze.py to not include any csv file handling

AdaraPutri committed:
added csv_analyze.py to independently analyze csv content as both main file or supporting file

AdaraPutri committed:
modified project_analysis.py to redirect csv files

AdaraPutri committed:
updated alt_analyze and its test to also use db connection, so tests do not fail

AdaraPutri committed:
fixed code flow so csv analysis is printed

AdaraPutri committed:
added tests for the csv module

AdaraPutri commented:
@taoTimTim @ivonanicetin 
Hey Timmi and Ivona, thanks for the thorough review!

The CSV handling wasn’t originally meant to be part of the main text or code analysis flow, as we established during our modification of the system architecture diagram that we only wanted to proceed with text (docx, txt, md, pdf) and code files as they were the main priority in the new requirements set by the TA and professor. 

This PR was originally only meant to quickly address one of the reviews from last week's PR mentioning that taking in a csv as supporting file for text main files that had data collection involved is "nice-to-have", as it allows more insight on the strength or weakness of the project process. 

However, I do understand the concern of pulling useful data for the resume from any files including csvs. Since integrating this in the text module would be bad practice of modularity, I added a separate csv_analyze.py that can both analyze supporting csv files (as part of a main text file) and analyze individual csv inputs. The llm in this script is only used for dataset summary while everything else related to the csv metadata is calculated using Pandas and Numpy.

The terminal output now includes:
* Number of rows, columns, headers, and data types
* Missing row counts and percentages
* Numeric column summaries (min, max, mean)
* User-typed or LLM-generated dataset summary
* Growth trend tracking for multi-version datasets (e.g., data.csv, data2.csv, data3.csv)

I have also refactored the flow in helpers.py so that it takes the extension from the DB instead of filepath (along with editing other scripts that were dependent on this). The addition of this module automatically solves the system exit error when csv is uploaded. I also added tests to ensure the csv module works properly.

Please test them and let me know if there are any changes to be made. I will likely polish the terminal output next week so it's resume ready, but please verify if the metrics for csv is what we need. Feel free to use these folders for testing: [single csv](https://drive.google.com/file/d/1Q37VlB71O5p_5U3TPCpiKSqJlMqM8dMk/view?usp=drive_link), [text file project with csv as supporting file](https://drive.google.com/file/d/15WdHzuzbAHHAeeA9H9NB_rSs1T2KNtBr/view?usp=drive_link)

I will update the PR description to include screenshots of the expected output.

taoTimTim reviewed (changes requested):
This is great, Adara! Sorry for the long PR review before, this one will be a bit shorter.

The metrics returned via the LLM are way better, more informative. Really good work here.
<img width="701" height="686" alt="image" src="https://github.com/user-attachments/assets/a0403e21-355f-4281-bee0-a864b5acf56c" />

Getting the filepath from the database is much better and has less duplication in our system.  Thanks for changing that.

However, two of the tests are failing:
<img width="1583" height="97" alt="image" src="https://github.com/user-attachments/assets/8b3d64cc-020c-41bd-8cee-3ce1d756f31d" />

They look to mostly be assertion errors, so most likely the tests need to be changed to match the implementation changes you made (or the functions are returning the wrong things).

Once these tests are running this can be merged! The metrics and LLM analysis are genuinely cool to see from a user's perspective.

AdaraPutri committed:
added normalization so tests run in both windows and mac

johaneshp reviewed (approved):
Thanks for working on this adara, this is a very good feature to have. all tests passed at my end, csv files are analyzed properly. This independent csv analysis allow the more detailed csv analysis.

taoTimTim reviewed (approved):
The tests are passing now! Thanks for doing all this and changing your implementation as requested. I know it was a lot more work than expected so I appreciate you following through.

Everything looks good now. Approved.

AdaraPutri committed:
Merge remote-tracking branch 'origin/main' into llm-text-analysis

AdaraPutri committed:
updated tests from main branch to pass with merge

ammaarkhan reviewed (approved):
Thanks for working on the additional tests and solving the issues mentioned by others. All tests are passing and working as intended. Great work!
<img width="1078" height="594" alt="Image" src="https://github.com/user-attachments/assets/4c303324-c5af-4bd1-9020-f714e86bcde2" />

ivonanicetin reviewed (changes requested):
Perfect! Thank you for the detailed reply to my previous review. And all the changes you have made since then. The original issue I talked about in the last review is fixed and now it doesn't even allow the user to accidentally chose a csv file as a main one which is smart. All the llm analysis looks great ! Great work.

However I am getting two tests that are failing. 

<img width="1247" height="207" alt="Screenshot 2025-11-09 190322" src="https://github.com/user-attachments/assets/2d02eef8-7634-4d19-9330-b322625de3b1" />

AdaraPutri commented:
@ivonanicetin Thanks for the review Ivona! That seems to be the same issue that Timmi faced, may I know if you are working on a macOS or windows? If it's the latter you might have to do 'git pull' to get the latest change (I added a filepath normalizer)

AdaraPutri committed:
reapply normpath to test file

ivonanicetin reviewed (approved):
I was on windows but your latest change fixed the issue, all the pytests are now passing. Great work!
